docs/proposals/env-vars.md

```diff
@@ -0,0 +1,516 @@
+
+The order of precedence is:
+
+- `DisallowedEnvironmentVariableKeys` in the `BuildStrategySpec` is considered first.  Any use of those keys in 
+  `BuildStep`, `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, `BuildRunSpec.[]EnvironmentVariable` results in a validation error.
+- `BuildStep` from the `BuildStrategySpec` is next.  The key/value pair from these override anything from
+  `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, `BuildRunSpec.[]EnvironmentVariable`
+- `BuildRunSpec.[]EnvironmentVariable` is next
+- followed by `BuildSpec.[]EnvironmentVariable`
+- and lastly, `BuildStrategySpec.[]EnvironmentVariable`
+
+Or conversely, the environment variables are applied in this fashion
+- first from `BuildStrategySpec.[]EnvironmentVariable` assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used
+- then from `BuildSpec.[]EnvironmentVariable`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
+  where values are replaced from any entries previously seeded
+- then from `BuildRunSpec.[]EnvironmentVariable`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
+  where values are replaced from any entries previously seeded
+- then from `BuildStrategySpec.[]BuildStep`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
+  where values are replaced from any entries previously seeded  
+
+Lastly, some yaml to complete the visualization.
+
+`BuildStrategy` with a disallow entry, a spec level default, and a final, always set via a `BuildStep`:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: ClusterBuildStrategy
+metadata:
+  name: something-that-uses-docker
+spec:
+  # nobody is allowed to change DOCKER_TLS_VERIFY
+  disallowedEnvironmentVariableKeys:
+    - 'DOCKER_TLS_VERIFY'
+  environmentVariables:
+    # for convenience, we add the version most our users can use, but we'll allow you to override
+    - name: DOCKER_API_VERSION
+      value: '1.19'
+  buildSteps:
+    - command:
+        - /usr/local/bin/docker
+        - build
+      env:
+        - name: DOCKER_CONFIG
+          value: /tekton/home/.docker
+
+```
+
+`Build` with an override of `DOCKER_API_VERSION`:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: something-that-uses-docker-build
+spec:
+  source:
+    url: https://github.com/myorg/myrepo
+  strategy:
+    name: something-that-uses-docker
+    kind: ClusterBuildStrategy
+  environmentVariables:
+    - name: DOCKER_API_VERION
+      value: '1.20'
+```
+
+`BuildRun` with an override of `DOCKER_API_VERSION`:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata:
+  name: something-that-uses-docker-buildrun
+spec:
+  environmentVariables:
+    - name: DOCKER_API_VERSION
```

Comment:
Is this the only environment variable the Build/BuildRun is allowed to set in this example ?

Reply:
nope per our exchange below